components/lock.rst

```diff
@@ -0,0 +1,258 @@
+.. index::
+   single: Lock
+   single: Components; Lock
+
+The Lock Component
+==================
+
+    The Lock Component creates and manages `locks`_, a mechanism to provide
+    exclusive access to a shared resource.
+
+.. versionadded:: 3.4
+    The Lock component was introduced in Symfony 3.4.
+
+Installation
+------------
+
+You can install the component in 2 different ways:
+
+* :doc:`Install it via Composer </components/using_components>` (``symfony/lock`` on `Packagist`_);
+* Use the official Git repository (https://github.com/symfony/lock).
+
+.. include:: /components/require_autoload.rst.inc
+
+Usage
+-----
+
+Locks are used to guarantee exclusive access to some shared resource. In
+Symfony applications, you can use locks for example to ensure that a command is
+not executed more than once at the same time (on the same or different servers).
+
+In order to manage the state of locks, a ``Store`` needs to be created first
+and then use the :class:`Symfony\\Component\\Lock\\Factory` class to actually
+create the lock for some resource::
+
+    use Symfony\Component\Lock\Factory;
+    use Symfony\Component\Lock\Store\SemaphoreStore;
+
+    $store = new SemaphoreStore();
+    $factory = new Factory($store);
+
+Then, a call to the :method:`Symfony\\Component\\Lock\\LockInterface::acquire`
+method will try to acquire the lock. Its first argument is an arbitrary string
+that represents the locked resource::
+
+    // ...
+    $lock = $factory->createLock('pdf-invoice-generation');
+
+    if ($lock->acquire()) {
+        // The resource "pdf-invoice-generation" is locked.
```

Comment:
Since a lock is an unmanaged resource, there is an exeception safety issue right here. Despite this is not a bug in the lock component itself, you should promote exception safety in examples IMO.

It can be fixed with a simple try / finally

```php
$lock = $factory->createLock('pdf-invoice-generation');
if ($lock->acquire()) {
    try {
        // ...
    } finally {
        $lock->release();
    }
}
```

You can even add an helper method to avoid this boilerplate

```php
$lock = $factory->createLock('pdf-invoice-generation');

$lock->with(function() {
   // ... 
});
```

With the following implementation

```php
public function with($callback)
{
    if (!$this->acquire()) {
        return false;
    }

    try {
        $callback();
    } finally {
        $this->release();
    }

    return true;
}
```

Reply:
You're right, this chapter was covered in `Blocking Locks` but it's better to promote it everywhere.